spec/actions_spec.rb

```diff
@@ -0,0 +1,179 @@
+      include Redcord::Base
+
+      attribute :value, T.nilable(Integer)
+
+      def self.name
+        'RedcordSpecModel'
+      end
+    end
+  end
+
+  let!(:migrator) do
+    Class.new do
+      extend Redcord::Migration::TTL
+    end
+  end
+
+  context 'create' do
+    it '#create!' do
+      instance = klass.create!(value: 3)
+      another_instance = klass.find(instance.id)
+      expect(instance.value).to eq another_instance.value
+    end
+
+    it 'validates types' do
+      # TODO: this should raise an error
+      klass.create!(value: '1')
+    end
+
+    it 'errors when id overflows a 64 bit signed integer' do
+      Redcord::Base.redis.set("#{klass.model_key}:id_seq", 2**63 - 2)
+
+      instance = klass.create!(value: nil)
+      expect(instance.id).to eq(2**63 - 1)
+
+      expect {
+        klass.create!(value: nil)
+      }.to raise_error(Redis::CommandError)
+    end
+  end
+
+  context 'update' do
+    it '#update!' do
+      instance = klass.create!(value: 3)
+      expect(instance.value).to eq 3
+
+      expect {
+        instance.value = '4'
+      }.to raise_error(TypeError)
+
+      expect {
+        instance.update!(value: '4')
+      }.to raise_error(TypeError)
+
+      instance.update!(value: 4)
+      expect(instance.value).to eq 4
+
+      another_instance = klass.find(instance.id)
+      expect(instance.value).to eq another_instance.value
+
+      instance = klass.new(value: 3)
+      instance.update!(value: 4)
+      another_instance = klass.find(instance.id)
+      expect(instance.value).to eq another_instance.value
+    end
+
+    it '#save!' do
+      instance = klass.new(value: 3)
+      expect(instance.value).to eq 3
+      expect(instance.updated_at).to be_nil
+      expect(instance.id).to be_nil
+      expect {
+        # instance_key is not available when id is nil
+        instance.instance_key
+      }.to raise_error(TypeError)
+      instance.save!
+
+      expect(instance.created_at).not_to be_nil
+      expect(instance.updated_at).not_to be_nil
+
+      another_instance = klass.find(instance.id)
+      expect(instance.value).to eq another_instance.value
+    end
+
+    it 'doesn\'t update redis until save!/update! is called' do
+      instance = klass.create!(value: 3)
+      instance.value = 4
+
+      another_instance = klass.find(instance.id)
+      expect(another_instance.value).to eq 3
+
+      instance.save!
+      another_instance = klass.find(instance.id)
+      expect(another_instance.value).to eq instance.value
+    end
+
+    it 'resets ttl' do
+      instance = klass.create!(value: 3)
+
+      klass.ttl(2.days)
+      migrator.change_ttl_passive(klass)
+      expect(klass.redis.ttl(instance.instance_key)).to eq(-1)
+
+      instance.save!
+      expect(klass.redis.ttl(instance.instance_key) > 0).to be true
+
+      klass.ttl(nil)
+      migrator.change_ttl_passive(klass)
+      instance.update!(value: 4)
+      expect(klass.redis.ttl(instance.instance_key)).to eq(-1)
```

Comment:
`change_ttl_passive` should make a call to redis to change the ttl for the model right? Do we need to call `save!` and `update!` as well?

Reply:
ah, so `change_ttl_passive` only updates the ttl for new records (which have been touched recently). For the ttl to be applied, we also need to touch the record using save! / update! (or we can call change_ttl_active (to be implemented), which automatically touches all existing records)